qbft/commit.go

```diff
@@ -82,35 +88,31 @@ func CreateCommit(state *State, config IConfig, root [32]byte) (*SignedMessage,
 
 		Root: root,
 	}
-	sig, err := config.GetShareSigner().SignRoot(msg, types.QBFTSignatureType, state.Share.SharePubKey)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed signing commit msg")
-	}
-
-	signedMsg := &SignedMessage{
-		Signature: sig,
-		Signers:   []types.OperatorID{state.Share.OperatorID},
-		Message:   *msg,
-	}
-	return signedMsg, nil
+	return Sign(msg, state.CommitteeMember.OperatorID, config.GetOperatorSigner())
 }
 
 func baseCommitValidationIgnoreSignature(
-	signedCommit *SignedMessage,
+	signedCommit *types.SignedSSVMessage,
 	height Height,
 	operators []*types.Operator,
 ) error {
-	if signedCommit.Message.MsgType != CommitMsgType {
-		return errors.New("commit msg type is wrong")
-	}
-	if signedCommit.Message.Height != height {
-		return errors.New("wrong msg height")
-	}
 
 	if err := signedCommit.Validate(); err != nil {
 		return errors.Wrap(err, "signed commit invalid")
 	}
 
+	msg, err := DecodeMessage(signedCommit.SSVMessage.Data)
```

Comment:
@MatheusFranco99

Reply:
As we discussed, the main problem is that, in `qbft`, we care about the contents of `qbft.Message` but we must also store the signed `types.SignedSSVMessage` to later add them in justifications.

E.g., this can be seen in the following flow of calls:
`qbft.Controller.ProcessMessage` $\to$ `qbft.Instance.ProcessMsg` $\to$ `qbft.uponPrepare` $\to$ `messageContainer.AddFirstMessageForSignerAndRound(signedSSVMessage)`

(The `SignedSSVMessage` had to be propagated until inside of `uponPrepare`).

As I see, there are two easy ways of managing this dual requirement. We either:
1. Send `SignedSSVMessage` as an argument in all functions that need it (or that call a function that needs it) and do the decoding inside the function if necessary (which is the current approach).
2. Or, send both the `SignedSSVMessage` and `qbft.Message` as arguments.

(1) forces us to do several decoding operations but allows the functions to be "independent" and doesn't need an assumption on the correctness of the relation of the messages. (2) though saves us from decoding, sends redundant information as argument (since we could have decoded) and makes the function "assume" that the input messages are correct (in a sense that the `qbft.Message` is really derived from the signed one).

To try to overcome both solutions we could try:
3. Define a structure in `qbft` as:
```go
type ProcessingMessage struct {
  SignedMessage types.SignedSSVMessage
  QBFTMessage qbft.Message
}
```
This structure would be used as arguments of the functions and would be created only once per message (either at the msg validation, runner or controller level). Decoding would be done just once, but we would need to assume that the argument is correct (if we don't want to check the correctness of the decoding in each function).

4. Another possibility is to define an interface that would do something similar
```go
type ProcessingMessage interface {
  GetSignedMessage() types.SignedSSVMessage
  GetQBFTMessage() qbft.Message
}
```
This abstracts how this decoding is done in spec (but of course implementation would implement a singleton decoding).
(4) abstracts this responsibility while (3) provides a more "direct" and "material" view of what's happening in the code for the readers.

What is your opinion?

Reply:
IMO the approach with the struct that holds both signed and qbft message (maybe it can even hold post-consensus decoded or we can have a different struct for it.). is better than decodeing each time, and it gives a better coupling to the both types of data belonging to the same message than passing in function parameters.

Reply:
@y0sher Great! I also agree that this is our best option 👍

Reply:
https://github.com/ssvlabs/ssv-spec/pull/440